File: client/src/Root/RoForge/Providers/OpenAI.lua
-- OpenAI Chat Completions API adapter (tools / function calling).
-- Direct call from Studio: the Authorization header goes straight to
-- api.openai.com. It never touches the RoForge backend or any proxy.

local HttpService = game:GetService("HttpService")
local Http = require(script.Parent.Parent.Http)

local OpenAI = {}

OpenAI.name = "openai"
OpenAI.defaultModel = "gpt-4.1" -- edit in Settings to use a newer model id

local function renderHistory(history)
	local msgs = {}
	for _, item in ipairs(history) do
		if item.role == "user" then
			table.insert(msgs, { role = "user", content = item.text })
		elseif item.role == "assistant" then
			local msg = { role = "assistant", content = item.text or "" }
			if item.calls and #item.calls > 0 then
				local toolCalls = {}
				for _, call in ipairs(item.calls) do
					local argsJson = "{}"
					pcall(function()
						argsJson = HttpService:JSONEncode(call.args or {})
					end)
					table.insert(toolCalls, {
						id = call.id,
						type = "function",
						["function"] = {
							name = call.name,
							arguments = argsJson,
						},
					})
				end
				msg.tool_calls = toolCalls
			end
			table.insert(msgs, msg)
		else -- role == "tool"
			table.insert(msgs, {
				role = "tool",
				tool_call_id = item.id,
				content = item.result,
			})
		end
	end
	return msgs
end

local function renderTools(tools)
	local out = {}
	for _, t in ipairs(tools) do
		table.insert(out, {
			type = "function",
			["function"] = {
				name = t.name,
				description = t.description,
				parameters = t.input_schema,
			},
		})
	end
	return out
end

function OpenAI.chat(cfg, system, history, tools)
	local model = (cfg.Model and cfg.Model ~= "") and cfg.Model or OpenAI.defaultModel
	local headers = {
		["Authorization"] = "Bearer " .. cfg.ApiKey,
		["content-type"] = "application/json",
	}
	local msgs = renderHistory(history)
	if system and system ~= "" then
		table.insert(msgs, 1, { role = "system", content = system })
	end
	local body = {
		model = model,
		messages = msgs,
	}
	if tools and #tools > 0 then
		body.tools = renderTools(tools)
	end

	local resp = Http.postJson("https://api.openai.com/v1/chat/completions", headers, body)
	if not resp.Success or (resp.StatusCode and resp.StatusCode >= 400) then
		return nil, Http.errorText(resp)
	end
	local data, err = Http.decode(resp.Body)
	if err then
		return nil, err
	end

	local choice = (data.choices or {})[1]
	local msg = (choice and choice.message) or {}
	local text = if type(msg.content) == "string" then msg.content else ""

	local calls = {}
	for _, tc in ipairs(msg.tool_calls or {}) do
		local fn = tc["function"] or {}
		local args = {}
		pcall(function()
			local decoded = HttpService:JSONDecode(tostring(fn.arguments or "{}"))
			if type(decoded) == "table" then
				args = decoded
			end
		end)
		table.insert(calls, { id = tc.id, name = fn.name, args = args })
	end
	return { text = text, calls = calls }, nil
end

return OpenAI

File: cli/src/providers/openai.js
// OpenAI Chat Completions — streaming (SSE) with function calling.
// The API key goes directly to api.openai.com (or cfg.openaiBaseUrl).
import { createSSE } from "../util.js";
import { ProviderError } from "./anthropic.js";

// fetch with a per-attempt timeout and ONE automatic retry on network-level
// failures (UND_ERR_CONNECT_TIMEOUT etc.) — user-initiated aborts are never
// retried.
export async function fetchWithRetry(url, opts, { timeoutMs = 60000, retries = 1 } = {}) {
  let lastErr;
  for (let attempt = 0; attempt <= retries; attempt++) {
    if (opts.signal?.aborted) throw lastErr || new ProviderError("aborted");
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(new Error("timeout")), timeoutMs);
    const onAbort = () => controller.abort(opts.signal.reason);
    if (opts.signal) opts.signal.addEventListener("abort", onAbort, { once: true });
    try {
      return await fetch(url, { ...opts, signal: controller.signal });
    } catch (e) {
      lastErr = e;
      if (opts.signal?.aborted) break; // user Ctrl+C — don't retry
      const sig = `${e.cause?.code || ""} ${e.name} ${e.message}`;
      const isNetwork = /UND_ERR|ECONN|ETIMEDOUT|EAI_AGAIN|EPIPE|EHOSTUNREACH|ENOTFOUND|timeout|aborted/i.test(sig);
      if (!isNetwork || attempt === retries) break;
      await new Promise((r) => setTimeout(r, 1200 * (attempt + 1)));
    } finally {
      clearTimeout(timer);
      if (opts.signal) opts.signal.removeEventListener("abort", onAbort);
    }
  }
  throw lastErr;
}

export async function chatStream(cfg, params, events = {}) {
  const key = cfg.openaiKey;
  if (!key) throw new ProviderError("No OpenAI API key. Run `roforge login` or set OPENAI_API_KEY.");
  const msgs = renderHistory(params.messages);
  if (params.system) msgs.unshift({ role: "system", content: params.system });
  const body = {
    model: params.model,
    stream: true,
    messages: msgs,
  };
  if (params.tools && params.tools.length) body.tools = renderTools(params.tools);

  let res;
  try {
    res = await fetchWithRetry(`${cfg.openaiBaseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${key}`,
        "content-type": "application/json",
      },
      body: JSON.stringify(body),
      signal: params.signal,
    });
  } catch (e) {
    throw new ProviderError(
      `network error calling OpenAI: ${e.cause?.code || e.message} (retried once — if it persists, check your connection)`
    );
  }
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    // OpenRouter retires free slugs; the 404 body names the paid replacement
    if (res.status === 404 && /unavailable for free/i.test(text)) {
      const m = text.match(/use this slug instead:\s*([^\s",}]+)/);
      throw new ProviderError(
        `${params.model} was retired from the free tier. Paid slug: ${m ? m[1] : "(see error)"} — ` +
          "or pick a live free model: https://openrouter.ai/models?max_price=0 (then /model openrouter:<slug>)"
      );
    }
    throw new ProviderError(`OpenAI HTTP ${res.status}: ${text.slice(0, 400)}`);
  }

  let text = "";
  let stopReason = null;
  let usage = { input_tokens: 0, output_tokens: 0 };
  const toolCalls = [];

  const sse = createSSE((ev) => {
    const d = ev.data;
    if (d === "[DONE]") return;
    if (typeof d !== "object" || !d) return;
    if (d.usage) usage = { ...usage, ...d.usage };
    const choice = (d.choices || [])[0];
    if (!choice) return;
    const delta = choice.delta || {};
    if (choice.finish_reason) stopReason = choice.finish_reason;
    if (typeof delta.content === "string" && delta.content) {
      text += delta.content;
      events.onText && events.onText(delta.content);
    }
    if (delta.tool_calls) {
      for (const tc of delta.tool_calls) {
        const idx = tc.index ?? 0;
        if (!toolCalls[idx]) toolCalls[idx] = { id: "", name: "", input: {}, _json: "" };
        const rec = toolCalls[idx];
        if (tc.id) {
          if (!rec.id) events.onToolStart && events.onToolStart(tc.function?.name || "tool", tc.id);
          rec.id = tc.id;
        }
        if (tc.function && tc.function.name) rec.name = tc.function.name;
        if (tc.function && tc.function.arguments) rec._json += tc.function.arguments;
      }
    }
  });

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    sse.push(decoder.decode(value, { stream: true }));
  }
  sse.end();

  const finalCalls = toolCalls
    .filter(Boolean)
    .map((rec) => {
      try {
        rec.input = rec._json ? JSON.parse(rec._json) : {};
      } catch {
        rec.input = {};
      }
      delete rec._json;
      return { id: rec.id, name: rec.name, input: rec.input };
    });

  const normReason = stopReason === "tool_calls" ? "tool_use" : stopReason === "stop" ? "end_turn" : stopReason;
  return { text, toolCalls: finalCalls, stopReason: normReason, usage };
}

export function renderHistory(history) {
  const msgs = [];
  for (const item of history) {
    if (item.role === "user") {
      msgs.push({ role: "user", content: item.text });
    } else if (item.role === "assistant") {
      const msg = { role: "assistant", content: item.text || "" };
      if (item.calls && item.calls.length) {
        msg.tool_calls = item.calls.map((c) => ({
          id: c.id,
          type: "function",
          function: { name: c.name, arguments: JSON.stringify(c.args || c.input || {}) },
        }));
      }
      msgs.push(msg);
    } else if (item.role === "tool") {
      msgs.push({ role: "tool", tool_call_id: item.id, content: item.result });
      // OpenAI has no image in tool messages — attach a follow-up user
      // message with an image_url data URI when the tool returned an image.
      if (item.image && item.image.base64) {
        const uri = `data:${item.image.mediaType || "image/png"};base64,${item.image.base64}`;
        msgs.push({
          role: "user",
          content: [
            { type: "text", text: `Image captured by tool ${item.name} (rendered from the Studio viewport):` },
            { type: "image_url", image_url: { url: uri } },
          ],
        });
      }
    }
  }
  return msgs;
}

export function renderTools(tools) {
  return tools.map((t) => ({
    type: "function",
    function: {
      name: t.name,
      description: t.description,
      parameters: t.inputSchema || t.input_schema,
    },
  }));
}
